### DISCUSSION TITLE
In-process tracing between C++/Python

### DISCUSSION THREAD

**Discussion by lidavidm** [open-telemetry/community#734] [Community Discussions]
We have a C++ library ([Apache Arrow](https://arrow.apache.org)) and we're considering instrumenting it with OpenTelemetry. However, it's often used through its Python bindings, and users of those bindings themselves are using OpenTelemetry. We'd like to make sure spans on one side link up to spans on the other.

So far, we're modeling this as two separate processes - when C++ calls into Python, we save the span/trace ID and inject it into the Python context with some glue code on the Python/Cython side. The opposite direction isn't implemented, but given Python->C++ calls are very common, I'd rather have some way for the C++ context to be updated when the Python context is updated (i.e. when setting the active span) and not on every call into C++. 

My question here is
1. Have there been any thoughts on 'official' or 'standard' support for doing things like this, or at least a recommended approach?
2. Is the approach here reasonable, or are there better ways to achieve this?

**Comment by tigrannajaryan** [comment]
@lidavidm thanks for bringing this up. 

>So far, we're modeling this as two separate processes - when C++ calls into Python, we save the span/trace ID and inject it into the Python context with some glue code on the Python/Cython side.

Can you please link to what you already have implemented?

@open-telemetry/specs-approvers I don't remember us discussing mixed language scenarios. Does anyone have any insights on context propagation?

**Reply by lidavidm** [reply]
Sure. It's in this PR: https://github.com/apache/arrow/pull/10260

In `server_middleware.h` the `TracingServerMiddleware` object exposes getters for a trace/span ID. (I guess I misspoke: we don't expose the current span, but the span for the current RPC in that part of the library. We have a wrapper over gRPC, hence, we also want to include tracing.)

In `_flight.pyx` we pull those values if available and inject them into the Python-side context: https://github.com/apache/arrow/pull/10260/files#diff-0ed358f5d42920d7f94cc500791976a2c158c4d72f4a6b231393534b2d13683bR1662-R1672

I realize that given this is potentially an O(implementations^2) kind of issue, it's impractical to support this across all language combinations. But I think the case of C++ with Python (and perhaps some other language pairs) is worth considering.

**Reply by reyang** [reply]
One possible way is to use the OpenTelemetry Python API, and bind it to the OpenTelemetry C++ SDK.
This has been discussed in the OpenTelemetry C++ SIG but was not funded. Definitely something we should explore if there are sufficient interest / funding.

**Reply by lidavidm** [reply]
Interesting, thanks for both your responses. The C++ binding approach sounds ideal in terms of functionality, though I can also see it being an implementation/maintenance burden - perhaps it's something worth pursuing once everything reaches 1.0, if users need more complete integration between both languages.

**Reply by reyang** [reply]
I think this is the nature of C/C++ - many language runtimes might expect bindings / foreign-function-interface (e.g. Python FFI/ctypes, Java/JNI, .NET/PInvoke) and interoperability.

cc @open-telemetry/cpp-approvers as FYI

**Reply by Oberon00** [reply]
If we expect many bindings for C/C++ it might be wisest to add a C binding and base further bindings on that. E.g. .NET and Python can consume most C APIs without the need to write native code.

**Reply by reyang** [reply]
Yeah, the idea is to have the C++ API/SDK as the core, and provide plain old C API on top of it (we have to make sure that there is no C++ exception, etc. and most of these are already covered by the OpenTelemetry C++ SIG).

**Reply by pyohannes** [reply]
> One possible way is to use the OpenTelemetry Python API, and bind it to the OpenTelemetry C++ SDK.

Another way would be providing a context implementation that exposes both a Python and a C++ interface. Both [Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/context/__init__.py#L54) and [C++](https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/context/runtime_context.h#L145) support pluggable context implementations. Then Python and C++ code running in one process would share a single context implementation, thus no need to pass around span and trace ids.

While complete C/C++ bindings would give other languages the performance benefits of C++, I think sharing just the context requires much less effort and is a more realistic short term solution.

**Reply by reyang** [reply]
That's another possible way with its own pros and cons. The challenge will be the extensibility - e.g. if someone is using a custom sampler in Python, the spans from C++ might not follow the same rule.

**Reply by Oberon00** [reply]
@pyohannes 

> Another way would be providing a context implementation that exposes both a Python and a C++ interface

How would that work? The Span stored in the Context would still be incompatible between Python and C++.

**Reply by pyohannes** [reply]
> How would that work? The Span stored in the Context would still be incompatible between Python and C++.

True, `Span` would need to be somehow wrapped, so both languages can access it.

> I think sharing just the context requires much less effort and is a more realistic short term solution.

I'll take that back, after thinking a bit more about it. It's a fragile solution. And setting up _two_ pipelines in a single process (one for C++ and one for Python) sounds very cumbersome.

**Reply by maxgolov** [reply]
> The Span stored in the Context would still be incompatible between Python and C++.

I guess we could have normalized it (serialized it) into `w3c traceparent` string, passing - marshalling that string. Which should be standard across languages. Then unmarshalling it in C++. And reconstructing the context from that. Then creating `Span` in C++ with `remote` parent. The other types in context could have been a strongly-typed key-value pairs. I've done something like this before for Go and C# to C++. Even via pure FFI-C API.

This is example how this could have been done via flexible key-value pairs in pure C, header only. We use this approach while passing key-value pairs from C to C++ in our products. It's C89:
https://github.com/microsoft/cpp_client_telemetry/blob/f14f9ba603c7ad26d5502e8cc2fc1458f3b152b4/lib/include/public/mat.h#L167

In Python:
- generate bindings to C API via [CFFI](https://cffi.readthedocs.io/en/latest/)
- in OpenTelemetry C++ SDK - add a separate C-to-C++ parameter passing handler with `extern "C"`
- that handler function intakes the Python-to-C struct, extracts the `traceparent` + strongly typed key-value pairs
- reconstructs the context from that as `remote Span`, same as what would've happend if it was HTTP or gRPC context propagation.
